File: libs/API/model/information_request/information_request.py
from libs.utils.lazy_property import lazyprop
from libs.utils.unix_time_stamp import from_time_stamp



class RFIs(object):

    def __init__(self):
        self._rfis = list()

    def add_rfi(self, rfi):
        for i_rfi in self._rfis:
            if rfi.id == i_rfi.id:
                i_rfi.update(rfi)
                return
        self._rfis.append(rfi)

    def delete_rfi(self, rfi):
        if not isinstance(rfi, InformationRequest):
            raise TypeError("argument must have type RFI")
        for l_rfi in self._rfis:
            if l_rfi.id == rfi.id:
                self._rfis.remove(l_rfi)

    def get_latest(self):
        return self._rfis[-1]


class InformationRequest(dict):

    def __init__(self, **kwargs):
        super(InformationRequest, self).__init__(**kwargs)

    @lazyprop
    def dueDate(self):
        return from_time_stamp(self['dueDate'])

    @dueDate.setter
    def dueDate(self, value):
        self['dueDate'] = value

    @lazyprop
    def state(self):
        return self['state']

    @state.setter
    def state(self, value):
        if value:
            self['state'] = value

    @lazyprop
    def priority(self):
        return self['priority']

    @priority.setter
    def priority(self, value):
            self['priority'] = value

    @lazyprop
    def searchType(self):
        return self['searchType']

    @searchType.setter
    def searchType(self, value):
        if value:
            self['searchType'] = value

    @lazyprop
    def externalRequestNumber(self):
        return self['externalRequestNumber']

    @externalRequestNumber.setter
    def externalRequestNumber(self, value):
            self['externalRequestNumber'] = value

    @lazyprop
    def requestSource(self):
        return self['requestSource']

    @requestSource.setter
    def requestSource(self, value):
        if value:
            self['requestSource'] = value

    @lazyprop
    def distributionList(self):
        return self['distributionList']

    @distributionList.setter
    def distributionList(self, value):
            self['distributionList'] = value

    @lazyprop
    def taskCategories(self):
        return self['taskCategories']

    @taskCategories.setter
    def taskCategories(self, value):
            self['taskCategories'] = value

    @lazyprop
    def subject(self):
        return self['subject']

    @subject.setter
    def subject(self, value):
            self['subject'] = value

    @lazyprop
    def description(self):
        return self['description']

    @description.setter
    def description(self, value):
            self['description'] = value

    @lazyprop
    def goals(self):
        return self['goals']

    @goals.setter
    def goals(self, value):
            self['goals'] = value

    @lazyprop
    def originalDocument(self):
        return self['originalDocument']

    @originalDocument.setter
    def originalDocument(self, value):
            self['originalDocument'] = value

    @lazyprop
    def approvedCopy(self):
        return self['approvedCopy']

    @approvedCopy.setter
    def approvedCopy(self, value):
            self['approvedCopy'] = value

    @lazyprop
    def createdDate(self):
        return from_time_stamp(self['createdDate'])

    @createdDate.setter
    def createdDate(self, value):
            self['createdDate'] = value

    @lazyprop
    def createdAt(self):
        return from_time_stamp(self['createdAt'])

    @createdAt.setter
    def createdAt(self, value):
            self['createdAt'] = value

    @lazyprop
    def createdBy(self):
        return self['createdBy']

    @createdBy.setter
    def createdBy(self, value):
            self['createdBy'] = value

    @lazyprop
    def id(self):
        return self['id']

    @id.setter
    def id(self, value):
            self['id'] = value

    @lazyprop
    def modifiedAt(self):
        return from_time_stamp(self['modifiedAt'])

    @modifiedAt.setter
    def modifiedAt(self, value):
            self['modifiedAt'] = value

    @lazyprop
    def modifiedBy(self):
        return self['modifiedBy']

    @modifiedBy.setter
    def modifiedBy(self, value):
            self['modifiedBy'] = value

    @lazyprop
    def internalRequestNumber(self):
        return self['internalRequestNumber']

    @internalRequestNumber.setter
    def internalRequestNumber(self, value):
            self['internalRequestNumber'] = value

    @lazyprop
    def previousRequests(self):
        return self['previousRequests']

    @previousRequests.setter
    def previousRequests(self, value):
            self['previousRequests'] = value

File: common/src/main/java/model/InformationRequest.java
package model;

import abs.TeelaEntity;
import data_for_entity.annotations.*;
import data_for_entity.data_providers.*;
import data_for_entity.data_types.FieldDataType;
import org.codehaus.jackson.annotate.JsonIgnore;
import org.codehaus.jackson.annotate.JsonIgnoreProperties;
import org.codehaus.jackson.annotate.JsonProperty;
import org.codehaus.jackson.map.annotate.JsonSerialize;

import java.text.SimpleDateFormat;
import java.util.*;

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonSerialize(include = JsonSerialize.Inclusion.NON_EMPTY)
public class InformationRequest extends TeelaEntity {
    
    @DataIgnore
    private String internalRequestNumber;
    private String externalRequestNumber;
    
    @DataProvider(RFIPrioritiesProvider.class)
    private Integer priority;
    @DataIgnore
    private String createdBy;
    @DataIgnore
    private String assignedTo;
    private String subject;
    private String description;
    private String goals;
    @WithFieldDataType(FieldDataType.DATE)
    private Date createdDate;
    @DataIgnore
    private Date lastRespondTime;
    @WithDataDependencies(provider = RFIDueDateProvider.class, fields = {"priority"})
    private Date dueDate;
    private String requestSource;
    @DataStatic("PENDING")
    private String state;
    @DataProvider(RFISearchTypeProvider.class)
    private InformationRequestSearchType searchType;
    @DataProvider(RFIDistributionProvider.class)
    private ArrayList<InformationRequestDistribution> distributionList = new ArrayList<>();
    @DataProvider(RFITaskCategoryProvider.class)
    private ArrayList<InformationRequestTaskCategory> taskCategories = new ArrayList<>();
    @DataIgnore
    private ArrayList<String> previousRequests;
    @DataIgnore
    private ArrayList<String> targets;
    @DataIgnore
    private FileAttachment approvedCopy;
    @DataIgnore
    private FileAttachment originalDocument;
    
    @DataIgnore
    @JsonIgnore
    public static SimpleDateFormat dateFormat = new SimpleDateFormat("dd.MM.yyyy HH:MM:");

    @JsonIgnore
    public FileAttachment getApprovedCopy() {
        return approvedCopy;
    }

    @JsonProperty
    public void setApprovedCopy(FileAttachment approvedCopy) {
        this.approvedCopy = approvedCopy;
    }

    @JsonIgnore
    public FileAttachment getOriginalDocument() {
        return originalDocument;
    }

    @JsonProperty
    public void setOriginalDocument(FileAttachment originalDocument) {
        this.originalDocument = originalDocument;
    }

    public void setSearchType(InformationRequestSearchType searchType) {
        this.searchType = searchType;
    }

    public InformationRequestSearchType getSearchType() {
        return searchType;
    }

    private Calendar getCalendar(){
        return Calendar.getInstance(TimeZone.getTimeZone("UTC"));
    }

    public String getInternalRequestNumber() {
        return internalRequestNumber;
    }

    public void setInternalRequestNumber(String internalRequestNumber) {
        this.internalRequestNumber = internalRequestNumber;
    }

    public String getExternalRequestNumber() {
        return externalRequestNumber;
    }

    public void setExternalRequestNumber(String externalRequestNumber) {
        this.externalRequestNumber = externalRequestNumber;
    }

    public Integer getPriority() {
        return priority;
    }

    @JsonProperty("priority")
    public void setPriority(Integer priority) {
        this.priority = priority;

    }

    public void setPriority(int priority) {
        this.priority = priority;
    }

    public String getCreatedBy() {
        return createdBy;
    }

    public void setCreatedBy(String createdBy) {
        this.createdBy = createdBy;
    }

    public String getAssignedTo() {
        return assignedTo;
    }

    public void setAssignedTo(String assignedTo) {
        this.assignedTo = assignedTo;
    }

    public String getSubject() {
        return subject;
    }

    public void setSubject(String subject) {
        this.subject = subject;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getGoals() {
        return goals;
    }

    public void setGoals(String goals) {
        this.goals = goals;
    }

    public Date getCreatedDate() {
        return createdDate;
    }

    public void setCreatedDate(Date createdDate) {
        this.createdDate = createdDate;
    }

    public Date getLastRespondTime() {
        return lastRespondTime;
    }

    public void setLastRespondTime(Date lastRespondTime) {
        this.lastRespondTime = lastRespondTime;
    }

    public Date getDueDate() {
        return dueDate;
    }

    public void setDueDate(Date dueDate) {
        this.dueDate = dueDate;
    }

    public String getRequestSource() {
        return requestSource;
    }

    public void setRequestSource(String requestSource) {
        this.requestSource = requestSource;
    }

    public String getState() {
        return state;
    }

    public void setState(String state) {
        this.state = state;
    }

    public List<InformationRequestDistribution> getDistributionList() {
        if (distributionList.size()== 0) {
            return null;
        }
        return distributionList;
    }

    public void setDistributionList(ArrayList<InformationRequestDistribution> distributionList) {
        this.distributionList = distributionList;
    }

    public List<InformationRequestTaskCategory> getTaskCategories() {
        if (taskCategories.size() == 0) {
            return null;
        }
        return taskCategories;
    }

    public void setTaskCategories(ArrayList<InformationRequestTaskCategory> taskCategories) {
        this.taskCategories = taskCategories;
    }

    public List<String> getPreviousRequests() {
        return previousRequests;
    }

    public void setPreviousRequests(ArrayList<String> previousRequests) {
        this.previousRequests = previousRequests;
    }

    public List<String> getTargets() {
        return targets;
    }

    public void setTargets(ArrayList<String> targets) {
        this.targets = targets;
    }

    
}